### DISCUSSION TITLE
Extending and modifying the functionality for copying PV names and pasting them into the Data Browser

### DISCUSSION THREAD

**Discussion by abrahamwolk** [ControlSystemStudio/phoebus#2972] [General]
Dear all,

At ESS we have a need for extending the functionality for copying PVs and pasting them into the Data Browser.

In particular, we have a need to copy multiple PVs from the Display Runtime, optionally annotating them with a label or comment, and pasting multiple PVs copied in this manner into the Data Browser with a single action. This would entail several changes to the existing functionality in Phoebus for copying PV names and pasting them into the Data Browser.

More concretely, we propose the following changes, and we would like to ask for feedback on these. This is of course only our current proposal, and we are happy to discuss this. I have categorised our proposed changes into the following three categories:

**1. Extended support for copying PV names by right-clicking in the Display Runtime**

We propose to extend the functionality available in the context-menu in the Display Runtime by introducing a sub-menu, which in turn contains the selectable functions:

 (1.A) "Copy PV Name to Clipboard" (the old "Copy PV to Clipboard")
 (1.B) "Copy PV Name to Clipboard with Description..." (new option)
 (1.C) "Append PV Name to Clipboard" (new option; appends the PV Name on a new line to the existing clipboard contents)
 (1.D) "Append PV Name to Clipboard with Description..." (new option)
 (1.E) "Copy PV Name to Clipboard with Value" (the old "Copy PV with Value to Clipboard")

I.e., the current context menu-items "Copy PV to Clipboard" and "Copy PV with Value to Clipboard" would be moved into this sub-menu.

Options (1.B) and (1.D) would open a modal dialog into which a description can be entered. The default description is the name of the current OPI. What would be pasted into the clipboard is then of the form `<PV Name>,<Description>` (i.e., the description is separated by a comma `,` from the PV name).

**2. Support for copying PV names in the Widget Info window**

We propose to add support in the "Widget Info" window (which is the same as "Display Info" window) for copying PV names in the display/widget as follows:
 - Add checkboxes to select PVs in the display/widget
 - Add button(s) with functionality similar to (1.A) to (1.D) described above.

**3. Modified and extended functionality to paste several PV names (with an optional display name) into the Data Browser**

We propose to modify and extend the functionality to paste several PV names into the Data Browser as follows:
 - Currently, several PV names can be dragged into the Data Browser using drag'n'drop. We propose to make it possible to also paste a set of PV names using a new option in the context menu of the Data Browser.
 - We propose to modify the format of the PV names that are pasted into the Data Browser as follows:
    1. The PV names are line separated (i.e., we propose to remove whitespace and other separators)
    2. Each PV name is optionally followed by a display name. In this case, the display name is separated from the PV name by a comma `,`. E.g.: `PV-NAME,This is a display name`. (The idea is that the description added by the functionality in (1.B) and (1.D) is parsed as a display name when pasted into the Data Browser.)

**Comment by kasemir** [comment]
Note that you can drag & drop PVs.
To start the drag from within the display runtime, you need to use control or the Mac command key to start the drag, then drop into the data browser.

If you need to add 3 PVs like that, it's pretty straight forward to control-drag the first, then the second, then the third.
Other ways for selecting the 3 PVs in the display runtime would likely create some new scheme that nobody will be able to remember.

**Reply by abrahamwolk** [reply]
The idea is that one can right-click on PV1, click "Copy PV Name". Then one can go to another PV (maybe on another display), right-click and select "Append PV Name", and so on. At any one point, only one PV is selected.

**Comment by kasemir** [comment]
Instead of "Append PV Name" on the source end I'd add a more convenient way to "add PV" to the target side.
Right now it's super easy to open a new data browser from one selected PV.
To then add more PVs, you can drag-drop more PVs to a data browser instance, but adding a copied one requires "Add PV", then pasting the copied name. Something like "Add PV from clipboard", or simply pre-populating the PV name in the "Add PV" dialog whenever the clipboard contains only one name might be a good idea, because that also works when you copied the one PV name from an email etc.

**Reply by abrahamwolk** [reply]
The copying of PV names into a list of PV names is not only for Phoebus: other applications can also take this format as input.

The idea is to add an option "Add PV(s) from the Clipboard" to the context menu of the Data Browser, which allows for the addition of one or several PV(s) from the clipboard. The axis can be set either individually for each PV or the added PVs can share an axis.

**Comment by abrahamwolk** [comment]
To make the proposal more concrete, here is a screenshot of the proposed changes to context menu of PVs in OPIs:
![contextMenu](https://github.com/ControlSystemStudio/phoebus/assets/122287276/19386b26-e5f7-403f-bbf4-22d08f0fca35)

When clicking on `Append PV Name to Clipboard with Description...`, the following dialog would be displayed:

![dialog](https://github.com/ControlSystemStudio/phoebus/assets/122287276/b4be9666-e046-4618-846a-aa2d3c120d50)

(The text of the `TextField` containing the description would be selected when the window opens, so that the description can be edited as easily as possible, and the dialog can be closed by pressing enter.)

Using the append functionality, one could, for example, have the following contents in the clipboard:
```
PV1,Process Variable 1
PV2,Process Variable 2
PV3,Process Variable 3
```

In the Data Browser, we propose the addition of the following context menu item:
![dataBrowser](https://github.com/ControlSystemStudio/phoebus/assets/122287276/4b143471-dfa5-4afe-9c09-142acc8e7a54)

which when clicked with the above contents in the clipboard, would open a generalized "Add PV(s)" dialog:
![addPVs](https://github.com/ControlSystemStudio/phoebus/assets/122287276/660b598b-c79a-45f6-ac7a-3b8e1cf64323)

**Proposed changes to the Widget Info dialog**
In addition to the above, we also propose to add a functionality to easily copy PV names (possibly with descriptions) from the "Widget Info" dialog.

**Comment by kasemir** [comment]
Would the "description" in the first dialog be pre-populated from the description in the PV metadata that's available for PV Access? What do you use for CA where that description is empty?
Is the same description then the default for the "Display Name" in the second dialog?

**Reply by abrahamwolk** [reply]
Currently, the design is that the default value is the Display Name of the OPI. (The motivation is that several similar PVs from similiar OPIs may be collected, and that the OPI name can meaningfully distinguish between the PVs in that case.)

**Reply by abrahamwolk** [reply]
Regarding your second question: yes, exactly. The "description" field is used for the "Display Name" field in the second dialog.

**Comment by georgweiss** [comment]
The sub-menu is bit overwhelming in my view. The description dialog could be triggered if user ticks a checkbox next to the sub-menu item text? This implies an additional click, but sub-menu would look cleaner.

**Reply by abrahamwolk** [reply]
I think that you're correct that there are many options now. However, I think that there are two drawbacks to a solution with checkboxes:
 1. As you wrote, more clicks are required.
 2. How to explain the functionality to the user: the UI is no longer "self-documenting" with checkboxes (unless descriptions are added to the labels, but that could make the labels quite long.)

**Comment by georgweiss** [comment]
How about a new top level context menu item: Append PV Name...

**Reply by abrahamwolk** [reply]
That could be an option.

**Comment by abrahamwolk** [comment]
Does anyone object to the basic functionality I have proposed here?
